### PR TITLE
Tor

### PR DESCRIPTION
**Pull request by nochiel**
This PR depends on <https://github.com/breez/breez/pull/139>.

It adds a "Tor" switch to Network Settings.

- [x] 20210908 Integrate guardianproject/tor-android.
- [x] 20210922 Integrate iCepa/Tor.Framework.
- [ ] Cleanup: 
    - [x] Incorporate previous code review from @yaslama.
    - [x] Make sure that I'm not using an extraneous imports in `Tor.java`.
    - [x] Clean up comments: `TODO` and `FIXME`

## UI
- [x] If Tor is enabled but fails to start, we disable Tor and show an error at app.Start.
- [ ] When Tor is enabled, should we show a Tor status icon on the main screen?
- ~Store TorConfig in SharedPreferences if `BreezMobile` uses them persistently. We don't need this setting in `breez` library.~
    - Unnecesary because TorConfig is needed by the server and not really used anywhere in the client.
- [x] Error handling:
    - If Tor is enabled but throws an error while starting or can't start at all, show the user an error with options to restart Breez with Tor disabled.
    - [x] Add option to [unexpected_error_dialog.dart](http://github.com/breez/breezmobile/blob/master/lib/routes/unexpected_error_dialog.dart) to disable Tor.

## Fix
- [x] Bug: Tor network setting doesn't correctly show that Tor is active/inactive.
- [x] Save Tor enabled/disabled state to user's settings.
- [x] Fixed: ~When using WebDav over Tor the backup is successful but the following occurs. I need to bisect to find out what changed in master because I have never encountered this error before.~
```log
I/GoLog   (29739): sendWebDavRequest: proxying request using tor with config {state:{NoUnkeyedLiterals:{} DoNotCompare:[] DoNotCopy:[] atomicMessageInfo:<nil>} sizeCache:0 unknownFields:[] Control:54598 Http:8118 Socks:9050}.
I/GoLog   (29739): 2021-09-08 14:27:52.696 [INF] CHANNELDB: creating shared channeldb service.
I/GoLog   (29739): 2021-09-08 14:27:52.704 [INF] CHDB: Checking for schema update: latest_version=20, db_version=20
I/GoLog   (29739): 2021-09-08 14:27:52.705 [INF] CHANNELDB: os.Remove(/data/user/0/com.breez.client/app_flutter/data/graph/mainnet/channel.db.old): remove /data/user/0/com.breez.client/app_flutter/data/graph/mainnet/channel.db.old: no such file or directory
I/GoLog   (29739): 2021-09-08 14:27:52.705 [INF] CHANNELDB: os.RemoveAll(/data/user/0/com.breez.client/app_flutter/bootstrap): <nil>
I/GoLog   (29739): 2021-09-08 14:27:52.705 [INF] CHANNELDB: channeldb was opened successfuly
I/GoLog   (29739): 2021-09-08 14:27:52.714 [INF] CHAIN: creating shared chain service.
I/GoLog   (29739): 2021-09-08 14:27:52.714 [ERR] CHAIN: peers error: database not open
I/GoLog   (29739): 2021-09-08 14:27:52.714 [ERR] DAEM: failed to create chainservice%!(EXTRA *errors.errorString=database not open)
I/GoLog   (29739): 2021-09-08 14:27:52.715 [INF] DAEM: Daemon.stop() called, stopping breez daemon alive=false
I/GoLog   (29739): 2021-09-08 14:27:52.715 [INF] DAEM: Daemon sent down event
```
- [x] Fixed: ~Regression: Verify that disabling Tor works consistently.~
    - ~An `unexpected error` screen appears after disabling Tor and restarting Breez.~

### CODE REVIEW

File: lib/bloc/account/account_bloc.dart [hunk 1]
```diff
@@ -3,6 +3,8 @@ import 'dart:convert';
 import 'dart:developer';
 import 'dart:io';
 
+import 'package:breez/bloc/tor/bloc.dart';
```
Comment (yaslama) [inline lib/bloc/account/account_bloc.dart:6]: This file seems to be missing

**Comment by tlindi** [comment]
> When Tor is enabled, should we show a Tor status icon on the main screen?
yes, please.

**Comment by tlindi** [comment]
Any priority for this issue? Would enable to use ones own nodes running Umbrel, which is very popular.

**Comment by kingonly** [comment]
Pending Tor v3 support in Neutrino

**Comment by tlindi** [comment]
ie This one?

https://github.com/lightningnetwork/lnd/issues/4973

**Comment by kingonly** [comment]
Yes. You can monitor the PR in Breez (not breezmobile).

**Comment by nochiel** [comment]
@ubbabeck @kingonly I'll merge this PR later today unless there are changes you'd like me to make.

**Comment by nochiel** [comment]
@ubbabeck @roeierez I've merged https://github.com/breez/breez/pull/139 and merged this PR against master.

So I think this PR is okay to merge with master now as the code hasn't changed since last year's review except to keep it up to date with master.

**Comment by erdemyerebasmaz** [comment]
@nochiel Hold on with the merge as @ubbabeck reported issues with routing yesterday.

**Comment by nochiel** [comment]
> @nochiel Hold on with the merge as @ubbabeck reported issues with routing yesterday.

Is there any update? Or details on what errors there were with routing? Is it an LND bug or …?

Also, wouldn't it be better for this to be merged so that the functionality can be kept in sync upstream instead of in a PR, seeing as the Tor functionality has been working since a year ago? 🙏🏾

**Comment by nochiel** [comment]
@kingonly Please could I get the rest of the bounty for this seeing as it's been working for over a year now and no changes have been requested? 

Happy holidays and thanks again!

**Comment by kingonly** [comment]
@nochiel I don't think it's fair. You agreed to receive half of the bounty because the task wasn't done. This was paid. We did the rest of the work. The lnd merge, which was brutal and the rest of the commits here:
https://github.com/breez/breez/pull/167/commits 
And:
https://github.com/breez/breezmobile/pull/1011/commits

**Comment by nochiel** [comment]
> @nochiel I don't think it's fair. You agreed to receive half of the bounty because the task wasn't done. This was paid. We did the rest of the work. The lnd merge, which was brutal and the rest of the commits here: https://github.com/breez/breez/pull/167/commits And: https://github.com/breez/breezmobile/pull/1011/commits

You literally did not communicate at all with me the whole time. I did everything to the spec we discussed. Smh. 😂 Have a good holiday, Roy.

**Comment by nochiel** [comment]
Hi 👋 guys,  
@roeierez @yaslama 

I'm not sure if this channel is still used, seeing that I've been removed from Breez Dev telegram and slack.

I'd like to request that if there's anyone here who's close enough friends with Roy, who can talk sense to him, that they perhaps consider doing so on my behalf with respect to the Tor bounty. I can't offer any incentive for you to intercede in this, other than proof of work.

In brief, I am asking if I can please be paid the remaining half of the Tor bounty.

Why? Because I did the work as we discussed and at no point in the past year did Roy or anyone else ask for more features or functionality or changes. That whole time I was still maintaining the PR, making sure it was synced with master and running.

Instead, Roy assigned my work, which was already done, to someone else and closed my PR.

I'm not interested in bickering in some weird back and forth about claiming who said what and when. Such arguments can't be resolved with dignity in any aspect of life because each person feels justified in their own perspective. In such an argument, Roy will claim that "you agreed to X", then I'll retort, "No, we discussed Y". Slack logs will be flung about, acrimony will flow, retorts will abound, ad infinitum.

However, I do feel I have been mistreated, disrespected, and that my time has been wasted. Roy has been unfair and acted in monstrously bad faith and clearly does not value my time or work. (Yet, he seems to think that only he has a claim to fairness. At such points it's useful to ask "What if I were this other person who I'm treating this way? Would it be fair?")

If Roy has a friend he listens to here, simply tell him to do the right thing. If he has an ethical backbone and good judgement, and I'd like to believe he does possess such things, then he'll know what that right thing is without having to be told what it is.

Have a good holiday all. 🥳 And good luck in your endeavours.

It's unfortunate that I won't get to contribute more to what you guys are building. I dig the Breez project, mission, and the vision to build Bitcoin. 🚀

Regardless of the outcome, I'll get over my bitterness eventually. 🖖🏿 I have no ill will towards Breez, the company or its people.

NB. If you feel compelled to tell me why I am wrong and deserve nothing, I urge you to refrain. It would waste both our times.😁

**Pull request closed without merging**